### PR TITLE
Integrate with Segment Anything Model (SAM)

### PR DESCRIPTION
**Pull request by wkentaro**
- Close https://github.com/wkentaro/labelme/discussions/1260
- Close https://github.com/wkentaro/labelme/discussions/1257

## TODO

- [x] Positive/negative click to run AI-assisted polygon annotation
- [x] Fix weird behavior when annotating many images https://github.com/wkentaro/labelme/pull/1262#issuecomment-1511834423
- [x] Support portrait images https://github.com/wkentaro/labelme/pull/1262#issuecomment-1521299774
- [x] Support 1-click annotation
- [x] Support double_click="close"
- [x] Real-time feedback without click https://github.com/wkentaro/labelme/pull/1265
- [x] Dropdown menu to change AI models

## Installation

1. Install conda
```
# Linux/Mac: command to install anaconda3 and labelme
curl -L https://github.com/wkentaro/dotfiles/raw/main/local/bin/install_anaconda3.sh | bash -s .
source .anaconda3/bin/activate

# Windows: check https://www.continuum.io/downloads
```

2. Install Labelme with `sam` branch.

```
git clone https://github.com/wkentaro/labelme.git
cd labelme
git checkout sam
pip install -e .
```

## Usage

```
labelme examples/instance_segmentation/data_annotated
```

1. Select "Create AI-Polygon"
2. Do `click` for the region you want to include
3. Do `shift+click` for the region you want to exclude
4. Do `ctrl/cmd+click` or `ctrl/cmd+shift+click` or double-click to finish
5. Press `esc` to go back and continue clicking

https://user-images.githubusercontent.com/4310419/235450790-e70e0351-21eb-4e2b-a14a-c874bb2bae1d.mp4

---

**NOTE: The response for the first annotation is SLOW (~20s)** because the AI model needs to extract image embedding from the entire image, but it is cached in the second you annotate.

### CODE REVIEW

**Comment by wkentaro** [comment]
@Johnny-liqiang Yes, I'm using ONNX exported model of Segment-Anything: https://github.com/facebookresearch/segment-anything#onnx-export

This is the forked repo for download and export scripts: https://github.com/wkentaro/segment-anything/tree/labelme

**Comment by healthonrails** [comment]
Thank you for the excellent work! I have a question about image embedding caching. Is it necessary for each image to have its own embedding?  Additionally, upon clicking, the mask's appearance is not displayed. It would be beneficial to highlight the mask before saving it. 

https://user-images.githubusercontent.com/1077233/232570096-5b9be2f3-a815-46ed-8f91-bba7e27d7984.mp4

The polygons in the third image may not be accurate, as demonstrated in the example video. This could be due to a caching issue with the image embedding.

**Comment by wkentaro** [comment]
@healthonrails 

> Is it necessary for each image to have its own embedding? 

Yes, and the image embedding cache is per image, so the third annotation is just failed by SAM (or can you confirm by only annotating the third one?).

**Comment by healthonrails** [comment]
If the _image_embedding has already been calculated for the previous image and is not None, it is likely that the program will reuse it instead of recalculating it for a new image.

 ``` def get_image_embedding(self):
        if self._image_embedding is None:
            with self._lock:
                self._image_embedding = compute_image_embedding(
                    image_size=self._image_size,
                    encoder_session=self._encoder_session,
                    image=self._image,
                )
                if len(self._image_embedding_cache) > 10:
                    self._image_embedding_cache.popitem(last=False)
                self._image_embedding_cache[
                    self._image.tobytes()
                ] = self._image_embedding
        return self._image_embedding
```

**Comment by wkentaro** [comment]
@healthonrails Thanks for checking! That sounds like something is wrong with cache hitting. I'll check that.

**Comment by not-a-swiss-conductor** [comment]
I use this formal command for test, **labelme examples/instance_segmentation/data_annotated**, but it can't work!!! 
I create a new environment to develop it, python = 3.10, numpy = 1.24, onnxruntime =  1.14. 
**NOTICE: python need >=3.9, for onnxruntime and numpy compatibility.**
```
Exception in thread Thread-1 (get_image_embedding):
Traceback (most recent call last):
  File "C:\Users\ysccc\anaconda3\envs\labelme\lib\threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "C:\Users\ysccc\anaconda3\envs\labelme\lib\threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "d:\code\labelme-sam\labelme\ai\models\segment_anything.py", line 55, in get_image_embedding
    self._image_embedding = compute_image_embedding(
  File "d:\code\labelme-sam\labelme\ai\models\segment_anything.py", line 107, in compute_image_embedding
    output = encoder_session.run(output_names=None, input_feed={"x": x})
  File "C:\Users\ysccc\anaconda3\envs\labelme\lib\site-packages\onnxruntime\capi\onnxruntime_inference_collection.py", line 196, in run
    raise ValueError("Model requires {} inputs. Input Feed contains {}".format(num_required_inputs, num_inputs))
ValueError: Model requires 6 inputs. Input Feed contains 1
```

**Comment by Johnny-liqiang** [comment]
> 

Are you running on Mac or Windows?, for me the Window, this Labelme with SAM doesn't work as well, it doesn't show creat AI Polygon this function when I right-clicked!

I believe the following install command only works for **Mac or Linux Systems**:

> git clone https://github.com/wkentaro/labelme.git
> cd labelme
> 
> # Install anaconda3 and labelme
> curl -L https://github.com/wkentaro/dotfiles/raw/main/local/bin/install_anaconda3.sh | bash -s .
> source .anaconda3/bin/activate
> pip install -e .


If you want to test SAM with Labelme on **Windows**:
SAM model 91M ViT-B SAM 508M ViT-L or 636M ViT-H have been integrated into Anylabeling tool: https://github.com/vietanhdev/anylabeling

I have installed it on Windows as well! very easy to use:

> Install Anaconda
> then conda activate any label environment
> then just run anylabeling


The best model I would say from SAM is this ViT-H with 636M numbers of parameters. It basically segments most the objects well! the other two models from SAM are not performing as well as this!

**Comment by healthonrails** [comment]
You need to switch to the "sam" branch and then proceed to install it.




> > 
> 
> @not-a-swiss-conductor Are you running on Mac or Windows?, for me the Window, this Labelme with SAM doesn't work as well, it doesn't show creat AI Polygon this function when I right-clicked!
> 
> I believe the following install command only works for **Mac or Linux Systems**:
> 
> > git clone https://github.com/wkentaro/labelme.git
> > cd labelme
> > curl -L https://github.com/wkentaro/dotfiles/raw/main/local/bin/install_anaconda3.sh | bash -s .
> > source .anaconda3/bin/activate
> > pip install -e .
> 
> If you want to test SAM with Labelme on **Windows**: SAM model 91M ViT-B SAM 508M ViT-L or 636M ViT-H have been integrated into [Anylabeling tool](https://github.com/vietanhdev/anylabeling):
> 
> I have installed it on Windows as well! very easy to use:
> 
> > Install Anaconda
> > then conda activate any label environment
> > then just run anylabeling
> 
> The best model I would say from SAM is this ViT-H with 636M numbers of parameters. It basically segments most the objects well! the other two models from SAM are not performing as well as this!

File: labelme/widgets/canvas.py [hunk 1]
```diff
@@ -383,16 +411,40 @@ def mousePressEvent(self, ev):
                         self.line[0] = self.current[-1]
                         if int(ev.modifiers()) == QtCore.Qt.ControlModifier:
                             self.finalise()
+                    elif self.createMode == "ai_polygon":
```
Comment (d710055071) [inline labelme/widgets/canvas.py:414]: I think this is a bug, which will make the last point the point of the positive sample. It should be modified as follows.

`                        if ev.modifiers() & QtCore.Qt.ControlModifier:
                            self.finalise()
                        else:
                            self.current.addPoint(
                                self.line.points[1],
                                label=self.line.point_labels[1],
                            )
                            self.line.points[0] = self.current.points[-1]
                            self.line.point_labels[0] = self.current.point_labels[
                                -1
                            ]`

File: labelme/ai/models/segment_anything.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    return np.linalg.norm(contour_end - contour_start, axis=1).sum()
+
+
+def compute_polygon_from_points(
+    image_size, decoder_session, image, image_embedding, points, point_labels
+):
+    input_point = np.array(points, dtype=np.float32)
+    input_label = np.array(point_labels, dtype=np.int32)
+
+    onnx_coord = np.concatenate([input_point, np.array([[0.0, 0.0]])], axis=0)[
```
[the remark this answers is not in the document]
Comment (wkentaro) [reply labelme/ai/models/segment_anything.py:125]: I feel the same, but it is from the official example. I haven't found time to investigate actual why.

https://github.com/facebookresearch/segment-anything/blob/main/notebooks/onnx_model_example.ipynb

image

**Comment by wkentaro** [comment]
@not-a-swiss-conductor @healthonrails
If you install Anaconda from https://www.continuum.io/downloads, it should also work on Windows. (I updated the instruction)

File: labelme/ai/models/segment_anything.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (d710055071) [reply labelme/ai/models/segment_anything.py:125]: Thank you very much for your answer

**Comment by luzhenyv** [comment]
I just tried the new release on Win10 with Python 3.10 and wanted to share my feedback.

While labeling an instance, it was difficult to identify what a mask or Polygon generated by prompt points should look like. The [SAM Demo](https://segment-anything.com/demo) renders a mask in real-time after adding inclusive/exclusive points, which makes segmentation more straightforward and user-friendly.

In the alpha version, I had to press `ctrl/cmd+click` or `ctrl/cmd+shift+click` to see the AI-polygon. If unsatisfied, I had to press `ESC` back to add inclusive/exclusive points and loop through the process repeatedly. It would be helpful to see the polygon in real-time as we label each instance, just like in the official demo. This feature would enable more accurate and efficient labeling.

**Comment by ZJDATY** [comment]
> @Johnny-liqiang Yes, I'm using ONNX exported model of Segment-Anything: https://github.com/facebookresearch/segment-anything#onnx-export
> 
> This is the forked repo for download and export scripts: https://github.com/wkentaro/segment-anything/tree/labelme

When I test, I will report a file not found error. May I ask if the exported model is stored in this path?

```
An error has occurred using proxy: {}
HTTPSConnectionPool(host='github.com', port=443): Max retries exceeded with url: /wkentaro/labelme/releases/download/sam-20230416/sam_vit_l_0b3195.quantized.encoder.onnx (Caused by ProxyError('Cannot connect to proxy.', OSError(0, 'Error')))
Traceback (most recent call last):
  File "D:\ProgramData\Anaconda3\envs\yolov8\lib\shutil.py", line 566, in move
    os.rename(src, real_dst)
FileNotFoundError: [WinError 2] 系统找不到指定的文件。: 'C:\\Users\\Administrator\\.cache/gdown\\tmppoquox3p\\dl' -> 'C:\\Users\\Administrator\\.cache/gdown\\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\pyworkspaces\labelme\labelme\app.py", line 373, in <lambda>
    lambda: self.toggleDrawMode(False, createMode="ai_polygon"),
  File "d:\pyworkspaces\labelme\labelme\app.py", line 977, in toggleDrawMode
    self._ai_model.set_image(utils.img_data_to_arr(self.imageData))
  File "d:\pyworkspaces\labelme\labelme\app.py", line 835, in _ai_model
    self._ai_model_initialized = ai.SegmentAnythingModel()
  File "d:\pyworkspaces\labelme\labelme\ai\models\segment_anything.py", line 28, in __init__
    md5="080004dc9992724d360a49399d1ee24b",
  File "D:\ProgramData\Anaconda3\envs\yolov8\Lib\site-packages\gdown\cached_download.py", line 117, in cached_download
    shutil.move(temp_path, path)
  File "D:\ProgramData\Anaconda3\envs\yolov8\lib\shutil.py", line 580, in move
    copy_function(src, real_dst)
  File "D:\ProgramData\Anaconda3\envs\yolov8\lib\shutil.py", line 266, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "D:\ProgramData\Anaconda3\envs\yolov8\lib\shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\Administrator\\.cache/gdown\\tmppoquox3p\\dl'
```


Can you provide instructions for exporting the model, or can you provide instructions for exporting the finished model?

File: labelme/widgets/canvas.py (hunk 1 above)
Comment (wkentaro) [reply labelme/widgets/canvas.py:414]: If you use `Ctrl/Cmd+Shift+Click`, it will be negative sample to close the polygon, so it should be okay?

**Comment by wkentaro** [comment]
> @healthonrails Thanks for checking! That sounds like something is wrong with cache hitting. I'll check that.

https://github.com/wkentaro/labelme/pull/1262/commits/83e957e7d60b9378cf4111ed4c30259346b44c4a should fix this.

**Comment by ZJDATY** [comment]
> * Close [segment anything support #1260](https://github.com/wkentaro/labelme/discussions/1260)
> * Close [Implement SAM #1257](https://github.com/wkentaro/labelme/discussions/1257)
> 
> ## TODO
> * [x]  Positive/negative click to run AI-assisted polygon annotation
> * [x]  Fix weird behavior when annotating many images [Integrate with Segment Anything Model (SAM) #1262 (comment)](https://github.com/wkentaro/labelme/pull/1262#issuecomment-1511834423)
> * [x]  Support portrait images [Integrate with Segment Anything Model (SAM) #1262 (comment)](https://github.com/wkentaro/labelme/pull/1262#issuecomment-1521299774)
> * [x]  Support 1-click annotation
> * [x]  Support double_click="close"
> * [ ]  Real-time feedback without click [Sam realtime render #1265](https://github.com/wkentaro/labelme/pull/1265)
> 
> ## Installation
> 1. Install conda
> 
> ```
> # Linux/Mac: command to install anaconda3 and labelme
> curl -L https://github.com/wkentaro/dotfiles/raw/main/local/bin/install_anaconda3.sh | bash -s .
> source .anaconda3/bin/activate
> 
> # Windows: check https://www.continuum.io/downloads
> ```
> 
> 2. Install Labelme with `sam` branch.
> 
> ```
> git clone https://github.com/wkentaro/labelme.git
> cd labelme
> git checkout sam
> pip install -e .
> ```
> 
> ## Usage
> ```
> labelme examples/instance_segmentation/data_annotated
> ```
> 
> 1. Select "Create AI-Polygon"
> 2. Do `click` for the region you want to include
> 3. Do `shift+click` for the region you want to exclude
> 4. Do `ctrl/cmd+click` or `ctrl/cmd+shift+click` to finish
> 5. Press `esc` to go back and continue clicking
> 
>  Untitled.8lehvfep.mp4 
> **NOTE: The response for the first annotation is SLOW (~20s)** because the AI model needs to extract image embedding from the entire image, but it is cached in the second you annotate.

Hello, thank you very much. I am using the largest model, but it looks very poor and completely different from what I saw on the official website. I hope to see the effect in real time after marking the points, otherwise I don't know how many points to click for the best effect, or I can add a toggle option to display the effect by pressing Ctrl all the time, or directly display the effect after clicking the points.

**Comment by wkentaro** [comment]
@f-macfarlane95 The SAM model can definitely handle splitted objects, but we need to support it with group_id correctly in Labelme. Let's add the feature of supporting splitted object annotation later.

**Comment by FreeSoaring** [comment]
I encountered an error：
(labelme) PS C:\Code\labelme> labelme examples/instance_segmentation/data_annotated
[INFO   ] __init__:get_config:71 - Loading config file from: C:\Users\Administrator\.labelmerc
Computing MD5: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx
MD5 matches: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx
Computing MD5: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.decoder.onnx
MD5 matches: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.decoder.onnx
Traceback (most recent call last):
  File "c:\code\labelme\labelme\app.py", line 372, in <lambda>
    lambda: self.toggleDrawMode(False, createMode="ai_polygon"),
  File "c:\code\labelme\labelme\app.py", line 969, in toggleDrawMode
    self.canvas.createMode = createMode
  File "c:\code\labelme\labelme\widgets\canvas.py", line 131, in createMode
    self._ai_model = labelme.ai.SegmentAnythingModel()
  File "c:\code\labelme\labelme\ai\models\segment_anything.py", line 35, in __init__
    self._encoder_session = onnxruntime.InferenceSession(encoder_path)
  File "C:\Users\Administrator\.conda\envs\labelme\lib\site-packages\onnxruntime\capi\onnxruntime_inference_collection.py", line 335, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "C:\Users\Administrator\.conda\envs\labelme\lib\site-packages\onnxruntime\capi\onnxruntime_inference_collection.py", line 368, in _create_inference_session
    sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx failed:D:\a\_work\1\s\onnxruntime\core/graph/model_load_utils.h:57 onnxruntime::model_load_utils::ValidateOpsetForDomain ONNX Runtime only *guarantees* support for models stamped with official released onnx opset versions. Opset 17 is under development and support for this is limited. The operator schemas and or other functionality may change before next ONNX release and in this case ONNX Runtime will not guarantee backward compatibility. Current official support for domain ai.onnx is till opset 15.
(labelme) PS C:\Code\labelme>

**Comment by d710055071** [comment]
hello，
Do you have any plans to add rectangular boxes?

File: labelme/widgets/canvas.py (hunk 1 above)
Comment (d710055071) [reply labelme/widgets/canvas.py:418]: If I don't use Ctrl+leftClick in the target, I will add a negative sample point as a positive sample point to 'self. current' and end up with an error. We don't know where the user will click

**Comment by luzhenyv** [comment]
> I encountered an error： (labelme) PS C:\Code\labelme> labelme examples/instance_segmentation/data_annotated [INFO ] **init**:get_config:71 - Loading config file from: C:\Users\Administrator.labelmerc Computing MD5: C:\Users\Administrator.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx MD5 matches: C:\Users\Administrator.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx Computing MD5: C:\Users\Administrator.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.decoder.onnx MD5 matches: C:\Users\Administrator.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.decoder.onnx Traceback (most recent call last): File "c:\code\labelme\labelme\app.py", line 372, in lambda: self.toggleDrawMode(False, createMode="ai_polygon"), File "c:\code\labelme\labelme\app.py", line 969, in toggleDrawMode self.canvas.createMode = createMode File "c:\code\labelme\labelme\widgets\canvas.py", line 131, in createMode self._ai_model = labelme.ai.SegmentAnythingModel() File "c:\code\labelme\labelme\ai\models\segment_anything.py", line 35, in **init** self._encoder_session = onnxruntime.InferenceSession(encoder_path) File "C:\Users\Administrator.conda\envs\labelme\lib\site-packages\onnxruntime\capi\onnxruntime_inference_collection.py", line 335, in **init** self._create_inference_session(providers, provider_options, disabled_optimizers) File "C:\Users\Administrator.conda\envs\labelme\lib\site-packages\onnxruntime\capi\onnxruntime_inference_collection.py", line 368, in _create_inference_session sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model) onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from C:\Users\Administrator.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx failed:D:\a_work\1\s\onnxruntime\core/graph/model_load_utils.h:57 onnxruntime::model_load_utils::ValidateOpsetForDomain ONNX Runtime only _guarantees_ support for models stamped with official released onnx opset versions. Opset 17 is under development and support for this is limited. The operator schemas and or other functionality may change before next ONNX release and in this case ONNX Runtime will not guarantee backward compatibility. Current official support for domain ai.onnx is till opset 15. (labelme) PS C:\Code\labelme>

You can try to upgrade your ONNX Runtime version. I am using 1.14.1, which runs smoothly. Your current version seems to be 1.9 or 1.10. Upgrading to the latest version should help you avoid compatibility issues.

**Comment by FreeSoaring** [comment]
Thank you for your help. I was able to solve the previous error using your method, which was caused by the low version of Python. It prevented this library from using the latest version. However, I got another error. What is the environment and Python version you are using?

``` err
(labelme) PS C:\Code\labelme> labelme examples/instance_segmentation/data_annotated
[INFO   ] __init__:get_config:70 - Loading config file from: C:\Users\Administrator\.labelmerc
..SegmentAnythingModel  __init__!!!!!!
Computing MD5: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx
MD5 matches: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx
Computing MD5: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.decoder.onnx
MD5 matches: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.decoder.onnx
Traceback (most recent call last):
  File "c:\code\labelme\labelme\app.py", line 372, in <lambda>
    lambda: self.toggleDrawMode(False, createMode="ai_polygon"),
  File "c:\code\labelme\labelme\app.py", line 969, in toggleDrawMode
    self.canvas.createMode = createMode
  File "c:\code\labelme\labelme\widgets\canvas.py", line 133, in createMode
    image=labelme.utils.img_qt_to_arr(self.pixmap.toImage())
  File "c:\code\labelme\labelme\utils\image.py", line 61, in img_qt_to_arr
    bytes_ = img_qt.bits().asstring(w * h * d // 8)
AttributeError: 'memoryview' object has no attribute 'asstring'
Traceback (most recent call last):
  File "c:\code\labelme\labelme\widgets\canvas.py", line 754, in paintEvent
    points = self._ai_model.predict_polygon_from_points(
  File "c:\code\labelme\labelme\ai\models\segment_anything.py", line 77, in predict_polygon_from_points
    image_embedding = self._get_image_embedding()
  File "c:\code\labelme\labelme\ai\models\segment_anything.py", line 70, in _get_image_embedding
    if self._thread is not None:
AttributeError: 'SegmentAnythingModel' object has no attribute '_thread'
QBackingStore::endPaint() called with active painter; did you forget to destroy it or call QPainter::end() on it?
QBackingStore::endPaint() called with active painter; did you forget to destroy it or call QPainter::end() on it?
QBackingStore::endPaint() called with active painter; did you forget to destroy it or call QPainter::end() on it?
QPainter::begin: Painter already active
(labelme) PS C:\Code\labelme> 

```

**Comment by FreeSoaring** [comment]
The first time I clicked 'Create AI-Polygon', I got this error message.

``` err 
(labelme) PS C:\Code\labelme> labelme examples/instance_segmentation/data_annotated
[INFO   ] __init__:get_config:70 - Loading config file from: C:\Users\Administrator\.labelmerc
..SegmentAnythingModel  __init__!!!!!!
Computing MD5: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx
MD5 matches: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.encoder.onnx
Computing MD5: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.decoder.onnx
MD5 matches: C:\Users\Administrator\.cache/gdown\https-COLON--SLASH--SLASH-github.com-SLASH-wkentaro-SLASH-labelme-SLASH-releases-SLASH-download-SLASH-sam-20230416-SLASH-sam_vit_l_0b3195.quantized.decoder.onnx
Traceback (most recent call last):
  File "c:\code\labelme\labelme\app.py", line 372, in <lambda>
    lambda: self.toggleDrawMode(False, createMode="ai_polygon"),
  File "c:\code\labelme\labelme\app.py", line 969, in toggleDrawMode
    self.canvas.createMode = createMode
  File "c:\code\labelme\labelme\widgets\canvas.py", line 133, in createMode
    image=labelme.utils.img_qt_to_arr(self.pixmap.toImage())
  File "c:\code\labelme\labelme\utils\image.py", line 61, in img_qt_to_arr
    bytes_ = img_qt.bits().asstring(w * h * d // 8)
AttributeError: 'memoryview' object has no attribute 'asstring'

```

AI did not work.

The program crashed with an error message when I tried to select an area on the image after clicking "Create AI-Polygon" for the second time.

``` err
Traceback (most recent call last):
  File "c:\code\labelme\labelme\widgets\canvas.py", line 754, in paintEvent
    points = self._ai_model.predict_polygon_from_points(
  File "c:\code\labelme\labelme\ai\models\segment_anything.py", line 77, in predict_polygon_from_points
    image_embedding = self._get_image_embedding()
  File "c:\code\labelme\labelme\ai\models\segment_anything.py", line 70, in _get_image_embedding
    if self._thread is not None:
AttributeError: 'SegmentAnythingModel' object has no attribute '_thread'
QBackingStore::endPaint() called with active painter; did you forget to destroy it or call QPainter::end() on it?
QBackingStore::endPaint() called with active painter; did you forget to destroy it or call QPainter::end() on it?
QPainter::begin: Painter already active
(labelme) PS C:\Code\labelme> 

```

**Comment by wkentaro** [comment]
Now it supports all 3 SAM models.

image

**Comment by Rusab** [comment]
Where can I get the SAM integrated Labelme?

**Comment by OceanWong1991** [comment]
@VicencPB Actually, The author already merge sam to master, and delete sam branch.
image

**Comment by OceanWong1991** [comment]
> Where can I get the SAM integrated Labelme?

@Rusab JUST Try the latest version!

**Comment by wkentaro** [comment]
@OceanWong1991 Thanks for answering the questions!

Yes, the latest version should include everything you need to use SAM.

@VicencPB @Rusab Please follow [the installation guide](https://github.com/labelmeai/labelme?tab=readme-ov-file#installation). To get the latest version, you need either of them:

1. install via pip: `pip install labelme --upgrade`
2. download the pre-built binary from [the releases](https://github.com/labelmeai/labelme/releases).

**Pull request merged**